### PR TITLE
Blocking I2C

### PR DESCRIPTION
**Pull request by hanhossain**
This takes the changes made in #26 and adds onto them. Tested with the SSD1306 (using the Heltec LoRa 32 board) and the MPU6050.

Thank you @fmckeogh for the groundwork!

### CODE REVIEW

**Comment by MabezDev** [comment]
Awesome! I'll take this out for a spin this weekend!

File: src/i2c.rs [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+            self.0.scl_start_hold.write(|w| w.time().bits(hold));
+            self.0.scl_stop_hold.write(|w| w.time().bits(hold));
+
+            // timeout
+            self.0.to.write(|w| w.time_out_reg().bits(tout.into()));
+        }
+    }
+
+    /// Helper function for determining which interface corresponds to the current instance
+    fn is_i2c0(&self) -> bool {
```
Comment (MabezDev) [inline src/i2c.rs]: Perhaps we could add a method to the `Instance` trait to determine which i2c instance this is? Similiar to the way the SPI Instance trait does it.

File: src/i2c.rs [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+                .scl_filter_cfg
+                .modify(|_, w| w.scl_filter_en().clear_bit()),
+        }
+    }
+
+    /// Sets the frequency of the I2C interface by calculating and applying the associated timings
+    fn set_frequency(&mut self, freq: u32) {
+        // i2c_hal_set_bus_timing(&(i2c_context[i2c_num].hal), freq, 1);
+        // i2c_ll_cal_bus_clk(80000000, freq, 0);
+        let half_cycle = ((80_000_000 / freq) / 2) as u16;
```
Comment (MabezDev) [inline src/i2c.rs]: I was surprised that this is actually a fixed constant, but after looking at the idf implementation it seems it is! Perhaps we should pull this out into a name `const`?

File: src/i2c.rs [hunk 3]
```diff
@@ -0,0 +1,638 @@
+use crate::gpio::{InputPin, InputSignal, OutputPin, OutputSignal};
+use crate::target::{i2c, DPORT, I2C0, I2C1};
+use core::{ops::Deref, ptr};
+
+const DPORT_BASE_ADDR: u32 = 0x3FF4_0000;
```
Comment (MabezDev) [inline src/i2c.rs:5]: This is the address of the `UART0`peripheral? Should this actually be `0x3FF0_0000` as per section 2.3.5 of tfm?

**Comment by hanhossain** [comment]
Hey @MabezDev, friendly bump :)

**Review by MabezDev — CHANGES_REQUESTED** [review]
@hanhossain my deepest apologies, I started this review and got side tracked by other things. I just had a few comments on the actual implementation. Unfortunately I have not yet had the chance to try this on actual hardware, but I will do this as soon as possible.

File: src/i2c.rs (hunk 1 above)
Comment (hanhossain) [reply src/i2c.rs]: Good call, I'll move this method into the trait

File: src/i2c.rs (hunk 3 above)
Comment (hanhossain) [reply src/i2c.rs:5]: No, there's a panic on startup when setting it as `0x3FF0_0000`:
```
rst:0x1 (POWERON_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0048,len:12
ho 0 tail 12 room 4
load:0x3fff0054,len:4800
load:0x40078000,len:17448
load:0x4007c428,len:4840
entry 0x4007c6a0
----- PANIC -----
PanicInfo { payload: Any, message: Some(called `Result::unwrap()` on an `Err` value: InvalidChipId(0)), location: Location { file: "examples/i2c.rs", line: 98, col: 30 } }
```

What's interesting is under §2.3.5, the notice says:

> Peripherals accessed by the CPU via 0x3FF40000 ~ 0x3FF7FFFF address space (DPORT address) can also be accessed via 0x60000000 ~ 0x6003FFFF (AHB address). (0x3FF40000 + n) address and (0x60000000 + n) address access the same content, where n = 0 ~ 0x3FFFF.

So I guess the DPORT register is at `0x3FF0_0000`; however, peripheral access starts at `0x3FF4_0000`.

File: src/i2c.rs (hunk 2 above)
Comment (hanhossain) [reply src/i2c.rs]: Refactored it into a `SOURCE_CLK_FREQ` const

**Comment by MabezDev** [comment]
Perfect! Thanks for the contribution @hanhossain!

**Pull request merged**